### DISCUSSION TITLE
std::map<std::string, std::string> generates a json array

### DISCUSSION THREAD

**Discussion by jflevesque** [nlohmann/json#3363] [Q&A] [answered]
`v3.7.0`

I am wondering if I facing a bug or an expected behavior. I have an `std::map<std::string, std::string>` that I am serializing as such:
``` cpp
std::string Serialize(const std::map<std::string, std::string>& mapToSerializeInJson)
{
    return nlohmann::json{mapToSerializeInJson}.dump();
}

void GivenAMap_WhenSerializing_ThenResultIsJsonObject()
{
    std::map<std::string, std::string> myMap;
    myMap.emplace("key1", "value1");
    myMap.emplace("key2", "value2");
    std::string actual = Serialize(myMap);
    EXPECT_EQ(R"({"key1": "value1","key2":"value2"}), actual);
}

```
Sadly, the expectation fails because I instead have an array: `[{"key1": "value1","key2":"value2"}]`.

I looked at past issues, but didn't notice any related to my issue. Any suggestion on how I can get the serialization to provide me with an object instead of an array?

**Comment by jflevesque** [comment]
To be honest, I am wondering if this method is not the culprit:
``` cpp
// check if each element is an array with two elements whose first
// element is a string
bool is_an_object = std::all_of(init.begin(), init.end(),
                                [](const detail::json_ref<basic_json>& element_ref)
{
    return element_ref->is_array() && element_ref->size() == 2 && (*element_ref)[0].is_string();
});
```
Why is it looking for `element_ref->is_array()` instead of `element_ref->is_object()`?

**Comment by gregmarr** [comment, marked as answer]
Do this instead (parens instead of braces):

    return nlohmann::json(mapToSerializeInJson).dump();

You're triggering the initializer list constructor instead of the map constructor.

**Reply by jflevesque** [reply]
Shouldn't the behavior be consistent between the 2 approaches?

**Reply by nlohmann** [reply]
No, because the library allows to construct from an initializer list.
